mesa/agent.py

```diff
@@ -21,3 +23,37 @@ def step(self):
     @property
     def random(self):
         return self.model.random
+
+    def as_json(self, filter: bool = True) -> str:
+        """Convert Agent attributes to JSON string.
+
+        Args:
+            filter: Whether to filter out unserializable objects and private attributes
+
+        Notes:
+            If an attribute is not JSON-serializable, it is replaced by its
+            string representation unless `filter` is set to True.
+        """
+
+        attributes = json.loads(
+            json.dumps(
+                self.__dict__, default=lambda a: "__REMOVE_ATR" if filter else str(a)
```

Comment:
Dumping and loading might not seem very efficient, but it is probably the most reliable way to find out which attributes can be serialized. (the attributes are later merged with the properties, hence the loading)